src/assertions.jl

```diff
@@ -0,0 +1,41 @@
+"""
+    Assertion
+
+An Assertion stores settings for checking the validity of a `AbstractArray` or `Tables.table` containing missing values.
+New assertions are expected to subtype `Impute.Assertion` and, at minimum,
+implement the `assert(data::AbstractVector{Union{T, Missing}}, ::<MyAssertion>)` method.
+"""
+abstract type Assertion end
+
+
+"""
+    assert(data, a::Assertion; kwargs...)
+
+If the assertion `a` fails then an `AssertionError` is thrown, otherwise the `data`
+provided is returned without mutation.
+
+# Keywords
+* `dims`: The dimension to apply the assertion along (e.g., observations dim)
+"""
+assert
+
+# A couple fallbacks for matrices and tables
+function assert(data::AbstractMatrix, a::Assertion; dims=1)
+    for var in varwise(data; dims=dims)
+        assert(var, a)
+    end
+    return data
+end
+
+function assert(table, a::Assertion)
+    istable(table) || throw(MethodError(a, (table, a)))
+    columntable = Tables.columns(table)
+
+    for cname in propertynames(columntable)
+        assert(getproperty(columntable, cname), a)
+    end
+
+    return table
+end
+
+include("assertions/threshold.jl")
```

Comment:
Would we want it here, or would it be nicer to put these all in the top file?
There isn't hundreds of files.

ChainRules puts them all in the top
https://github.com/JuliaDiff/ChainRules.jl/blob/e2cfd58b983a5023fd649b79fd980bd3428280b8/src/ChainRules.jl#L26-L45

as does ChainRulesCore
https://github.com/JuliaDiff/ChainRulesCore.jl/blob/43e2113ebf482b74ba292c17e5b3845be4f5faef/src/ChainRulesCore.jl#L15-L29

Reply:
I guess I prefer seeing the abstract type API first, but I don't have a strong opinion. What's the benefit to putting it earlier?

Reply:
I guess marginally 
easier to delete or rename the file if I know where it will be referenced.

Mostly just for constancy?

Reply:
Oh, n/m looks like julia doesn't like that because then the threshold code doesn't know about the `Assertion` type yet.